ivadomed/loader/loader.py

```diff
@@ -1121,3 +1053,91 @@ def __init__(self, bids_df, subject_file_lst, target_suffix, contrast_params, sl
 
         super().__init__(self.filename_pairs, slice_axis, cache, transform, slice_filter_fn, task, self.roi_params,
                          self.soft_gt, is_input_dropout)
+
+    def annotate_raters(self, target_suffix, target_filename, derivative):
+        for idx, suffix_list in enumerate(target_suffix):
+            # If suffix_list is a string, then only one rater annotation per class is available.
+            # Otherwise, multiple raters segmented the same class.
+            if isinstance(suffix_list, list):
+                for suffix in suffix_list:
+                    if suffix in derivative:
+                        target_filename[idx].append(derivative)
+            elif suffix_list in derivative:
+                target_filename[idx] = derivative
+            
+            return target_filename
```

Comment:
If I'm not mistaken, the return should be outside of the for loop.

This was not picked up by our tests, maybe we should add a test to make sure all target_suffix are considered, both when provided as a list and as a list of list.

Reply:
@mariehbourget  I am in the process of writing a unit test for this now, however I am bit confused what the intention is for `target_filename`? I noticed that we appear to be setting it as a list of filenames, but looking elsewhere in the code it seems like `target_filename` is supposed to be a string and not a list. Was the intention to return a list of target_filenames or just one? If we want to just return one file name then I think in the `elif` block we should probably return the filename as soon as we find a matching derivative in the `suffix_list`? In the `if` block it appears we are trying to return a list of filenames? I just want to make sure my assumptions here are correct before pushing the change with the unit test

Reply:
@mariehbourget  I am in the process of writing a unit test for this now, however I am bit confused what the intention is for `target_filename`? I noticed that we appear to be setting it as a list of filenames, but looking elsewhere in the code it seems like `target_filename` is supposed to be a string and not a list. Was the intention to return a list of target_filenames or just one? If we want to just return one file name then I think in the `elif` block we should probably return the filename as soon as we find a matching derivative in the `suffix_list`? In the `if` block it appears we are trying to return a list of filenames? I just want to make sure my assumptions here are correct before pushing the change with the unit test

Reply:
@gisellemartel 
The intention is to return a list of `target_filename` corresponding to the list of `target_suffix` of the config file.

There are 2 cases in the `elif` block:
1) When `target_suffix` is a list containing 1 string to train a single class model.
2) When `target_suffix` is a list containing more than 1 string to train a multi class model.

For example (from the data_testing dataset), we have the image `sub-unf1_T1w.nii.gz`. In the derivatives, we have 2 different labels for this image: `sub-unf01_T1w_lesion-manual.nii.gz` and  `sub-unf01_T1w_seg-manual.nii.gz`.

1) In case 1: the `target_suffix` list is `["_lesion-manual"]` to train a model the class `lesion` only. Then the `target_filename` should be `["sub-unf01_T1w_lesion-manual.nii.gz"]`.
2) In case 2: the `target_suffix` list is `["_lesion-manual", "_seg-manual"]` to train both the class `lesion` and the class `seg`. Then the `target_filename` should be `["sub-unf01_T1w_lesion-manual.nii.gz", "sub-unf01_T1w_seg-manual.nii.gz"]`. This is the case that was not detected by the unit testing earlier when the function was only returning the first match.

The `if isinstance(suffix_list, list):` block deals with a 3rd case where we have `target_suffix` as a list of list of string.
I don't think we have an example for this in the testing dataset or a unit test for that case.
Here the `target_suffix` would be something like `[["_lesion-manual-rater1", "_lesion-manual-rater2"]]` for the single class case and `[["_lesion-manual-rater1", "_lesion-manual-rater2"], ["_seg-manual-rater1", "_seg-manual-rater2"]]` for the multi class case (meaning than we have 2 different target for the class `lesion` and 2 different target for the class `seg`), and the `target_filename` would also be a list of list of string with the same form.

Reply:
Thank you for the very detailed explanation!! This was quite helpful. I have added a unit test for now that tests the simpler case when we have `target_suffix` that is a list. In the case of having a list of lists, this seems to break the code elsewhere before the modifications to the loader I made. I would suggest perhaps putting this in a separate issue. What do you think @mariehbourget?